File: src/lib/backend-client/annotations.ts
/**
 * Typed front-end client for the per-URL annotations commands.
 *
 * Why this file exists:
 * - PathKeep stores notes and tags against URLs in the canonical archive
 *   (migration 011 + `vault-core::annotations`). The Browse detail panel
 *   writes them; the future Recall surface reads them.
 * - Keeping the typed client here means routes never type raw command
 *   names ("set_url_notes") and shield against transport renames.
 *
 * Main declarations:
 * - `annotationsClient`
 *
 * Source-of-truth notes:
 * - Transport contract: `docs/architecture/desktop-command-surface.md`.
 * - Shape contract: `vault_core::models::annotations` (UrlAnnotation,
 *   SetNotesRequest, ReplaceTagsRequest).
 */

import { call } from './shared'

/**
 * Per-URL annotation bundle returned by the backend. `notes` is the empty
 * string when no note has been written; `tags` is an empty array when no
 * tag is attached.
 */
export interface UrlAnnotation {
  url: string
  notes: string
  tags: string[]
  updatedAt: string
  createdAt: string
  sourceProfile?: string | null
}

export interface SetNotesRequest {
  url: string
  notes: string
  sourceProfile?: string | null
}

export interface ReplaceTagsRequest {
  url: string
  tags: string[]
  sourceProfile?: string | null
}

export const annotationsClient = {
  getUrlAnnotation: (url: string) =>
    call<UrlAnnotation | null>('get_url_annotation', { url }),
  setUrlNotes: (request: SetNotesRequest) =>
    call<UrlAnnotation>('set_url_notes', { request }),
  replaceUrlTags: (request: ReplaceTagsRequest) =>
    call<UrlAnnotation>('replace_url_tags', { request }),
  listUrlAnnotations: (limit?: number) =>
    call<UrlAnnotation[]>('list_url_annotations', { limit: limit ?? null }),
  searchUrlAnnotations: (query: string, limit?: number) =>
    call<UrlAnnotation[]>('search_url_annotations', {
      query,
      limit: limit ?? null,
    }),
}

File: src-tauri/src/commands/annotations.rs
//! Tauri commands for per-URL annotations (notes + tags).
//!
//! These commands let the desktop shell write the notes textarea and tag
//! chips from the paper Browse detail panel into the canonical archive.
//! Reads use the session database key so encrypted archives stay walled
//! behind the App Lock session.

//! All five commands touch encrypted SQLite, so they run on the blocking thread pool: on the
//! 14.4M-record baseline an annotation read/write/search must never block the WebView thread.

#[cfg(not(test))]
use super::blocking::run_blocking_command;
#[cfg(not(test))]
use crate::{session::SessionState, worker_bridge};
#[cfg(not(test))]
use tauri::State;

#[cfg(not(test))]
#[tauri::command]
/// Returns the annotation bundle for a URL (None when nothing is written), off the UI thread.
pub(crate) async fn get_url_annotation(
    state: State<'_, SessionState>,
    url: String,
) -> Result<Option<vault_core::UrlAnnotation>, String> {
    let key = state.get_key();
    run_blocking_command("get_url_annotation", move || {
        worker_bridge::get_annotation_impl(key.as_deref(), &url)
    })
    .await
}

#[cfg(not(test))]
#[tauri::command]
/// Sets or clears the notes body for a URL. Empty body removes the row. Off the UI thread.
pub(crate) async fn set_url_notes(
    state: State<'_, SessionState>,
    request: vault_core::SetNotesRequest,
) -> Result<vault_core::UrlAnnotation, String> {
    let key = state.get_key();
    run_blocking_command("set_url_notes", move || {
        worker_bridge::set_notes_impl(key.as_deref(), request)
    })
    .await
}

#[cfg(not(test))]
#[tauri::command]
/// Replaces the full tag set for a URL. An empty list removes all tags. Off the UI thread.
pub(crate) async fn replace_url_tags(
    state: State<'_, SessionState>,
    request: vault_core::ReplaceTagsRequest,
) -> Result<vault_core::UrlAnnotation, String> {
    let key = state.get_key();
    run_blocking_command("replace_url_tags", move || {
        worker_bridge::replace_tags_impl(key.as_deref(), request)
    })
    .await
}

#[cfg(not(test))]
#[tauri::command]
/// Lists URLs that carry at least one annotation, newest first. Off the UI thread.
pub(crate) async fn list_url_annotations(
    state: State<'_, SessionState>,
    limit: Option<usize>,
) -> Result<Vec<vault_core::UrlAnnotation>, String> {
    let key = state.get_key();
    run_blocking_command("list_url_annotations", move || {
        worker_bridge::list_annotations_impl(key.as_deref(), limit)
    })
    .await
}

#[cfg(not(test))]
#[tauri::command]
/// Searches notes by case-insensitive substring across the archive, off the UI thread.
pub(crate) async fn search_url_annotations(
    state: State<'_, SessionState>,
    query: String,
    limit: Option<usize>,
) -> Result<Vec<vault_core::UrlAnnotation>, String> {
    let key = state.get_key();
    run_blocking_command("search_url_annotations", move || {
        worker_bridge::search_annotations_impl(key.as_deref(), &query, limit)
    })
    .await
}
